cirq-core/cirq/ops/arithmetic_operation.py

```diff
@@ -241,8 +244,217 @@ def _apply_unitary_(self, args: 'cirq.ApplyUnitaryArgs'):
+        or else a list of qubit/qudit dimensions. Registers that are set to a
+        classical constant must not be mutated by the arithmetic operation
+        (their value must remain fixed when passed to `apply`).
+
+        Registers are big endian. The first qubit is the most significant, the
+        last qubit is the 1s qubit, the before last qubit is the 2s qubit, etc.
+
+        Returns:
+            A list of constants and qubit groups that the operation will act
+            upon.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def with_registers(self: TSelfGate, *new_registers: Union[int, Sequence[int]]) -> TSelfGate:
+        """Returns the same operation targeting different registers.
+
+        Args:
+            new_registers: The new values that should be returned by the
+                `registers` method.
+
+        Returns:
+            An instance of the same kind of operation, but acting on different
+            registers.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def apply(self, *register_values: int) -> Union[int, Iterable[int]]:
+        """Returns the result of the operation operating on classical values.
+
+        For example, an addition takes two values (the target and the source),
+        adds the source into the target, then returns the target and source
+        as the new register values.
+
+        The `apply` method is permitted to be sloppy in three ways:
+
+        1. The `apply` method is permitted to return values that have more bits
+            than the registers they will be stored into. The extra bits are
+            simply dropped. For example, if the value 5 is returned for a 2
+            qubit register then 5 % 2**2 = 1 will be used instead. Negative
+            values are also permitted. For example, for a 3 qubit register the
+            value -2 becomes -2 % 2**3 = 6.
+        2. When the value of the last `k` registers is not changed by the
+            operation, the `apply` method is permitted to omit these values
+            from the result. That is to say, when the length of the output is
+            less than the length of the input, it is padded up to the intended
+            length by copying from the same position in the input.
+        3. When only the first register's value changes, the `apply` method is
+            permitted to return an `int` instead of a sequence of ints.
+
+        The `apply` method *must* be reversible. Otherwise the operation will
+        not be unitary, and incorrect behavior will result.
+
+        Examples:
+
+            A fully detailed adder:
+
+            ```
+            def apply(self, target, offset):
+                return (target + offset) % 2**len(self.target_register), offset
+            ```
+
+            The same adder, with less boilerplate due to the details being
+            handled by the `ArithmeticOperation` class:
```

Comment:
```suggestion
            handled by the `ArithmeticGate` class:
```

Also update other parts of the docstring to reflect that it's now a gate and not an operation.